### DISCUSSION TITLE
Best way to handle multiple incoming requests that trigger external APIs

### DISCUSSION THREAD

**Discussion by 0xjgv** [fastapi/fastapi#8703] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
import asyncio


app = FastAPI()

async def expensive_request():
    asyncio.sleep(10)
    return {"foo": "bar"}

@app.get("/request")
async def request():
    result = await expensive_request() 
    return {"message": result}
```

### Description

The `expensive_request` method calls an external API service that is rate limited. When multiple users call the `/request` endpoint at the same time, the `expensive_request` gets triggered several times. What are the ways to group these multiple requests into one awaited one?

### Operating System

Linux

### FastAPI Version

0.65.2

### Python Version

3.9.9

**Comment by kayx23** [comment]
@juangaitanv perhaps limit your fastapi rate as well with library like `slowapi`? You can limit the rate for the entire endpoint, per IP src etc., and add some front end handling.

**Comment by 0xjgv** [comment]
> @juangaitanv perhaps limit your fastapi rate as well with library like `slowapi`? You can limit the rate for the entire endpoint, per IP src etc., and add some front end handling.

Thank you for your suggestion @Kayx23! Do you know of any way to do it without rate limiting the customer?

**Comment by jgould22** [comment]
Another option is to have a downstream queue whose producer is your endpoint and whose consumer batches up some of the requests currently in the queue preforms expensive request

**Comment by raphaelauv** [comment]
If the client have a limiting setting use it :

example of aiohttp
```python
SIZE_POOL_AIOHTTP = 100
timeout = aiohttp.ClientTimeout(total=2)
connector = aiohttp.TCPConnector(family=AF_INET, limit_per_host=SIZE_POOL_AIOHTTP)
cls.aiohttp_client = aiohttp.ClientSession(timeout=timeout, connector=connector)
```

full example -> https://github.com/raphaelauv/fastAPI-aiohttp-example/blob/92506131237f2b0d75b97a9c988ea23e46a82848/src/fastAPI_aiohttp/fastAPI.py#L28


Or put a Semaphore around the operation you want limit 

```python
SIZE_POOL = 10
sem = asyncio.Semaphore(SIZE_POOL)

async def expensive_request():
    async with sem:
        asyncio.sleep(10)
        
```

**Comment by 0xjgv** [comment]
Thank you @raphaelauv! Do you know of any method to group pending futures (i.e. same pending requests wait for its response)?

**Comment by DavidSantacruzR** [comment]
Perhaps you might want to use node.js along with fastApi to handle many parallel HTTP requests. Have NodeJs to place multiple requests into a an event queue and from there use your expensive_request script.

**Comment by 0xjgv** [comment]
Thank you @DavidSantacruzR! Something like this -> https://github.com/Dunkelheit/pending-promise-recycler?
Do you know of any similar implementation in Python?

**Comment by DavidSantacruzR** [comment]
@juangaitanv well not exactly like that. In other words, you might want to use a child_process from the Node packages (https://nodejs.org/api/child_process.html) and using this to call the script where you have the function or the python implementation.

**Comment by 0xjgv** [comment]
One way, I found easy to implement, to wait for already pending requests (`expensive_request`) is to use a cache in the following way:

```python
import asyncio
from fastapi import FastAPI

EXPENSIVE_REQUEST_KEY = "expensive"
tasks_cache = {}

app = FastAPI()


async def coro_wrapper(result):
    return result


async def expensive_request():
    print("Starting Expensive request...")
    await asyncio.sleep(5)
    print("Finished Expensive request...")
    return {"foo": "bar"}


@app.get("/request")
async def request():
    global tasks_cache

    future = None
    if task := tasks_cache.get(EXPENSIVE_REQUEST_KEY):
        if task.done():
            future = coro_wrapper(task.result())
        else:
            future = task
    else:
        tasks_cache[EXPENSIVE_REQUEST_KEY] = asyncio.create_task(expensive_request())
        future = tasks_cache[EXPENSIVE_REQUEST_KEY]

    result = await future
    tasks_cache = {}  # Reset cache
    return {"message": result}
```

What are your thoughts on this implementation 🙏 ?
